### PR TITLE
[train] Add torchrec variant to the training ingest benchmark

### PR DESCRIPTION
**Pull request by justinvyu**
## Summary

Adds a `task=recsys` variant to the Train + Data ingest, which trains a [DLRM RecSys model](https://arxiv.org/abs/1906.00091) on the Criteo dataset using [TorchRec](https://pytorch.org/torchrec/), modeled after [this example](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm) which is also a [MLPerf benchmark](https://github.com/mlcommons/training/tree/master/recommendation_v2/torchrec_dlrm).
* The model takes in a bunch of dense (ex: time of day, age, etc.) and sparse (ex: user id, item id, category, etc.) features and predicts a probability of a click / engagement.
* The model is characterized by large embedding tables to support a huge number of users/items (ex: multiple embedding tables with 40M+ rows), which takes up a large amount of GPU memory. The model is also low compute, since the embedding lookup and fully connected layers are relatively cheap.

This PR also updates the Train + Data ingest benchmark scaffolding to allow more flexibility in the train loop, since torchrec has more customized model initialization (fusing optimizer step with the backward pass, wrapping in the torchrec `DistributedModelParallel` class, abstracting away the loss function into the model forward pass, etc.) and didn't fit into the original mold I created for the image classification training.

## 📊 Benchmark Results Summary

| Dataloader   | Training Step     | Global Throughput (rows/sec) | Training Step Avg Time (s) | Notes  |
|--------------|-------------------|-------------------------------|-----------------------------|---------------------------------------|
| Mock         | Skip Training      | **16,497,545**                | ~0.000075                   |           |
| Ray Data     | Skip Training      | **1,204,491** (including first batch pipeilne warmup time)<br/> **5,558,343** (excluding first batch pipeline warmup time)              | 0.0053                      | Train step time still includes H2D and sparse data dist shuffling.    |
| Mock         | Full Training      | **1,128,657**                 | 0.0517                      | Matches torchrun baseline. This row result combined with the mock + skip training result above show that training is the main bottleneck.             |
| Ray Data     | Full Training      | **660,225**   (including first batch pipeline warmup time)<br/>  **1,151,438** (excluding first batch pipeline warmup time)             | 0.0524                      | The ~70s pipeline warmup time is what drags down the avg global throughput, but otherwise the "steady state" throughput matches the mock dataloader + full training performance.     |

### Sanity check

I also ran the [original example training script](https://github.com/facebookresearch/dlrm/tree/main/torchrec_dlrm) with the same mock dataloader as a sanity check, expecting the 2nd row mock dataloader + full training performance to be comparable. [Find the experiment here.](https://gist.github.com/justinvyu/9a3d1a21ff6edf8b1023939a42317659) The experiment gave a throughput of `1,179,648 rows/s`.

## TODOs for followup PRs

* Use the full training dataset (~1TB). Right now I am using the validation set (~30GB) for both training and validation. This is due to some blockers I faced: fitting preprocessors took too long (`groupby().count()`).
* Add the synthetic multi-hot data augmentation step. (Is this actually needed?)
* Data shuffling.
* Check that model accuracy converges to the benchmark number after 1 epoch.
* Add profiling to check whether data is pipelined correctly and fix benchmark timers (use cuda events, synchronize every N steps).

### CODE REVIEW

File: release/train_tests/benchmark/recsys/torchrec_runner.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        # lr_scheduler = LRPolicyScheduler(
+        #     optimizer, args.lr_warmup_steps, args.lr_decay_start, args.lr_decay_steps
+        # )
+
+        self.pipeline = TrainPipelineSparseDist(
+            self.model, self.optimizer, device, execute_all_batches=True
+        )
+
+    def train_step(self, train_dataloader):
+        self.pipeline.progress(train_dataloader)
```
Comment (justinvyu) [inline release/train_tests/benchmark/recsys/torchrec_runner.py:40]: Torchrec does some things differently and didn't fit the old `TrainLoopRunner` mold.
1. Creates this `TrainPipelineSparseDist` that wraps the model and optimizer, so the user doesn't call `model(input)` or compute the loss function directly. This pipelines a few batches at a time (moving to GPU, doing collective communication, forward and backward pass).
2. It also doesn't call next on the data iterator directly, but rather does that in `pipeline.progress(dataloader)`.
3. Does *not* wrap the model in DDP, instead wraps it in a torchrec `DistributedModelParallel` utility class.

Comment (justinvyu) [reply release/train_tests/benchmark/recsys/torchrec_runner.py:40]: So, I pulled out some methods that can be defined by subclasses. `setup`, `train_step`, validate_step`. Setup initializes all model related components that should be used in the training/validate steps.

File: release/train_tests/benchmark/recsys/criteo.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                features[features == None] = ""
+                return {cat_feature: features}
+
+            value_counts = [
+                (group[cat_feature], group["count()"])
+                for group in ds.map_batches(filter_features)
+                .groupby(key=cat_feature)
+                .count()
+                .take_all()
+            ]
```
Comment (justinvyu) [inline release/train_tests/benchmark/recsys/criteo.py:178]: Right now I'm blocked on this taking too long on the full TB dataset, which is why I'm just setting the training set to be the smaller validation set for now. I precomputed these values and saved them as files on s3.

File: release/train_tests/benchmark/recsys/criteo.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+def get_ray_dataset(stage: str = "train"):
+    ctx = DataContext.get_current()
+    # Enable hash-shuffling
+    ctx.shuffle_strategy = ShuffleStrategy.HASH_SHUFFLE
+
+    stage_to_path = {
+        # TODO: The preprocessor fitting (groupby().count()) takes too long
+        # on the full training dataset, so just run with the valid dataset for now.
+        "train": "s3://ray-benchmark-data-internal/criteo/tsv.gz/valid",
```
Comment (justinvyu) [inline release/train_tests/benchmark/recsys/criteo.py:134]: See above

File: release/train_tests/benchmark/recsys/recsys_factory.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    def collate_fn(self, batch):
+        return convert_to_torchrec_batch_format(batch)
+
+
+class TorchRecConfig(BaseModel):
+    embedding_dim: int = 128
+    num_embeddings_per_feature: List[int] = [
+        40000000,
+        39060,
```
Comment (justinvyu) [inline release/train_tests/benchmark/recsys/recsys_factory.py:70]: Note: This model config may only fit on A100 GPU. Need to check A10

File: release/train_tests/benchmark/recsys/criteo.py (hunk 3 above)
Comment (srinathk10) [reply release/train_tests/benchmark/recsys/criteo.py:237]: I see. How big is the Train & Validation dataset?

Comment (justinvyu) [reply release/train_tests/benchmark/recsys/criteo.py:237]: Train dataset is around 1TB. Valid dataset is around 30GB.

File: release/train_tests/benchmark/recsys/criteo.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        batch: Dict[str, np.ndarray], mapping_ref: ray.ObjectRef
+    ):
+        mapping: Dict[str, int] = ray.get(mapping_ref)
+        for cat_feature in DEFAULT_CAT_NAMES:
+            batch[cat_feature] = np.vectorize(
+                lambda k: mapping.get(cat_feature, {}).get(k, LOW_FREQUENCY_INDEX)
+            )(batch[cat_feature])
+        return batch
+
+    ds = ds.map_batches(
```
Comment (raulchen) [inline release/train_tests/benchmark/recsys/criteo.py:310]: since we are already precomputing `categorical_values_to_index`.
Is it a good idea to also precompute `categorical_values_to_indices` and ` concat_and_normalize_dense_features`?

File: release/train_tests/benchmark/recsys/criteo.py (hunk 2 above)
Comment (raulchen) [reply release/train_tests/benchmark/recsys/criteo.py:281]: Do you think in practice people would want to do this on the fly, instead of precomputing it? 
It looks that there is no randomization, so probably not necessary to repeatedly do this for each epoch.

File: release/train_tests/benchmark/recsys/torchrec_runner.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    def validate_step(self, val_dataloader):
+        return 0
+
+    def _save_training_state(self, local_dir: str):
+        return
```
Comment (raulchen) [inline release/train_tests/benchmark/recsys/torchrec_runner.py:46]: why early return here? 
Can you add a comment if the following is not needed?

File: release/train_tests/benchmark/recsys/torchrec_runner.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        torch.save(self.model.state_dict(), os.path.join(local_dir, "model.pt"))
+        torch.save(self.optimizer.state_dict(), os.path.join(local_dir, "optimizer.pt"))
+
+    def _load_training_state(self, local_dir: str):
+        return
```
Comment (raulchen) [inline release/train_tests/benchmark/recsys/torchrec_runner.py:51]: ditto

File: release/train_tests/benchmark/recsys/torchrec_runner.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        )
+        # lr_scheduler = LRPolicyScheduler(
+        #     optimizer, args.lr_warmup_steps, args.lr_decay_start, args.lr_decay_steps
+        # )
+
+        self.pipeline = TrainPipelineSparseDist(
+            self.model, self.optimizer, device, execute_all_batches=True
+        )
+
+    def train_step(self, train_dataloader):
```
Comment (raulchen) [inline release/train_tests/benchmark/recsys/torchrec_runner.py:39]: handle the skip_train_step case?

**Review by raulchen — COMMENTED** [review]
mostly LGTM. left some small comments.

File: release/train_tests/benchmark/recsys/criteo.py (hunk 5 above)
Comment (justinvyu) [reply release/train_tests/benchmark/recsys/criteo.py:310]: I was thinking the `categorical_values_to_index` is the "preprocessor" that's being fit. The `concat_and_normalize_dense_features` is the online preprocessing that can be done so that we don't need to store the whole dataset again in this intermediate format.

File: release/train_tests/benchmark/recsys/torchrec_runner.py (hunk 6 above)
Comment (justinvyu) [reply release/train_tests/benchmark/recsys/torchrec_runner.py:46]: Oh oops, I will add in the model saving logic.

File: release/train_tests/benchmark/recsys/torchrec_runner.py (hunk 1 above)
Comment (justinvyu) [reply release/train_tests/benchmark/recsys/torchrec_runner.py]: I found a different API in torchrec that allows calling next on the data iterator directly and does the forward/backward/optimizer steps explicitly.

However, there are are still differences between the 2, so I kept the overridable `Runner` skeleton.

File: release/train_tests/benchmark/recsys/torchrec_runner.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            # ),
+        ]
+
+        pipeline = StagedTrainPipeline(pipeline_stages=pipeline)
+
+        def dataloader_with_torchrec_pipeline():
+            while batch := pipeline.progress(dataloader_iter):
+                yield batch
+
+            pipeline.flush_end()
```
Comment (justinvyu) [inline release/train_tests/benchmark/recsys/torchrec_runner.py:77]: This is the torchrec utility to pipeline the device transfer, data shuffling, and model forward/backward steps.

File: release/train_tests/benchmark/recsys/torchrec_runner.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    def _get_model_and_optim_filenames(self):
+        rank = ray.train.get_context().get_world_rank()
+        return f"model_shard_{rank=}.pt", f"optimizer_shard_{rank=}.pt"
+
+    def _save_training_state(self, local_dir: str):
+        # Save sharded model and optimizer state.
+        model_filename, optimizer_filename = self._get_model_and_optim_filenames()
+
+        torch.save(self.model.state_dict(), os.path.join(local_dir, model_filename))
```
Comment (justinvyu) [inline release/train_tests/benchmark/recsys/torchrec_runner.py:107]: Embedding table shards are on different GPUs, so we need to do distributed checkpointing. This checkpoint format must be loaded on the same number of workers and GPU types, since it was sharded with a specific plan.

File: release/train_tests/benchmark/runner.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        for batch in train_dataloader:
+            with self._metrics["train/step"].timer():
+                if not self.benchmark_config.skip_train_step:
+                    self._train_step(batch)
+                else:
+                    # There are still some GPU operations that do not involve the model
+                    # (ex: non_blocking batch.to(device), data shuffling collective ops)
+                    # that need to be synchronized.
+                    if torch.cuda.is_available():
+                        torch.cuda.synchronize()
```
Comment (justinvyu) [inline release/train_tests/benchmark/runner.py:152]: TODO for a future PR: measure time with cuda events and occasional cuda syncs.

File: release/train_tests/benchmark/recsys/torchrec_runner.py (hunk 10 above)
Comment (raulchen) [reply release/train_tests/benchmark/recsys/torchrec_runner.py:107]: let's also add this note to code comments? (same for others)

**Review by raulchen — APPROVED** [review]

File: release/ray_release/byod/byod_install_torchrec.sh [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+# shellcheck disable=SC2102
+
+set -exo pipefail
+
+pip3 install torch==2.6.0 \
+  torchrec==1.1.0 \
+  fbgemm-gpu==1.1.0 \
+  --extra-index-url https://download.pytorch.org/whl/cu121
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply release/ray_release/byod/byod_install_torchrec.sh:9]: I tried but ran into issues with torch and fbgemm-gpu incompatibility that would force me to upgrade torch from 2.3.0 to 2.6.0+ across the board which is too big of an effort and shouldn't block this PR.

**Review by aslonnie — APPROVED** [review]

File: release/train_tests/benchmark/recsys/recsys_factory.py
Comment (matthewdeng) [inline release/train_tests/benchmark/recsys/recsys_factory.py:1]: Just curious, any reason this is `recsys_factory` and not just more specifically `torchrec_factory` (to parallel `torchrec_runner`?

**Review by matthewdeng — APPROVED** [review]

File: release/release_tests.yaml [hunk 13]
```diff
@@ -2161,6 +2161,43 @@
...
+
+  # TODO: Keep these manual for now -- there are a still a few issues to resolve.
+  frequency: manual
+  team: ml
+
+  cluster:
+    byod:
+      type: gpu
+      post_build_script: byod_install_torchrec.sh
+    cluster_compute: compute_configs/compute_gpu_8xA100_aws.yaml
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply release/release_tests.yaml:2177]: It's actually been pretty easy to acquire lately.

The model size would need to be changed if we want to run on smaller GPUs. (e.g. reduce size of embedding tables) That might run into other issues since the dataset may not be compatible if a bunch of embedding indices get removed.

All the existing benchmarks are also run on A100 (mlperf, torch example readme, etc.).

File: release/release_tests.yaml [hunk 14]
```diff
@@ -2161,6 +2161,43 @@
...
+
+    - __suffix__: skip_training.ray_data
+      run:
+        timeout: 600
+        script: RAY_TRAIN_V2_ENABLED=1 python train_benchmark.py --task=recsys --dataloader_type=ray_data --num_workers=8 --train_batch_size=8192 --num_epochs=1 --skip_train_step --skip_validation_at_epoch_end
+
+    - __suffix__: full_training.mock_dataloader
+      run:
+        timeout: 600
+        script: RAY_TRAIN_V2_ENABLED=1 python train_benchmark.py --task=recsys --dataloader_type=mock --num_workers=8 --train_batch_size=8192 --num_epochs=1
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply release/release_tests.yaml:2193]: oh yeah I can match it.

File: release/train_tests/benchmark/recsys/recsys_factory.py [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+                ),
+                dense_in_features=len(DEFAULT_INT_NAMES),
+                dense_arch_layer_sizes=args.dense_arch_layer_sizes,
+                over_arch_layer_sizes=args.over_arch_layer_sizes,
+                dcn_num_layers=args.dcn_num_layers,
+                dcn_low_rank_dim=args.dcn_low_rank_dim,
+                dense_device=device,
+            )
+        elif args.interaction_type == "projection":
+            raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply release/train_tests/benchmark/recsys/recsys_factory.py:135]: The mlperf benchmark only uses original and DCN interaction types, so I just didn't bother including the imports for the stuff below. I'll just delete this code for sanity.
